lib/meta: add platform constraints algebra

Atemu opened the pull request:
Previously, we were able to express platform constraints in packages using e.g.

    meta.platforms = with lib.platforms; linux ++ darwin;

These platform checks only worked on a coarse level using lists of pre-defined platform strings. It was not possible to express that only e.g. 64bit platforms or that only GNU libcs are supported previously.

This introduces an algebra based on platform patterns which allows you to express arbitrarily complex platform constraints using simple expressions without needing to rely on `stdenv.*Platform.is*` bools.

Once this is in wide-spread use, it can be used for accurate rendering of platform support in manuals and search sites.

It also removes the need for dependency on a package's `stdenv` for expressing which platforms are broken as this can now be expressed using platform constraints instead. This makes progress of potentially evaluating a package's meta information independently of the package or its inputs.

## Things we still need to discuss:

* [ ] The actual integration in platforms/badPlatforms
* [ ] Performance testing (this is on the hot path!)
* [ ] How/when to coordinate this with https://search.nixos.org?
* [ ] Docs and guidelines

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Atemu committed:
lib.platform: init platform constraints algebra

Previously, we were able to express platform constraints in packages using e.g.

    meta.platforms = with lib.platforms; linux ++ darwin;

These platform checks only worked on a coarse level using lists of pre-defined
platform strings. It was not possible to express that only e.g. 64bit platforms
or that only GNU libcs are supported previously.

This introduces an algebra based on platform patterns which allows you to
express arbitrarily complex platform constraints using simple expressions
without needing to rely on `stdenv.*Platform.is*` bools.

Once this is in wide-spread use, it can be used for accurate rendering of
platform support in manuals and search sites.

It also removes the need for dependency on a package's `stdenv` for expressing
which platforms are broken as this can now be expressed using platform
constraints instead. This makes progress of potentially evaluating a package's
meta information independently of the package or its inputs.

Atemu committed:
lib.meta: introduce meta.platformConstraints

It would be trivially possible to support platformConstraints in meta.platforms
and meta.badPlatforms by simply replacing all occurrences of platformMatch with
evalConstraint but that would require wrapping the platform constraints inside
of a list when declaring them in packages which doesn't make a whole lot of
sense given that they have native support for combining multiple conditions.

It would be possible to allow lists or non-lists or even choose which method to
use based on whether the value passed is a list but I think it's simpler if we
just introduced a new meta attribute instead to make a clean cut between legacy
platform declarations and platform constraints declaration.

Atemu commented:
The commit I just pushed will be made redundant in a follow-up but I want to see ofBorg eval perf because @adisbladis suggested that even seemingly tiny changes can have an immense impact there because this is on the hot path for all packages.

Atemu commented:
They have also provided an alternative implementation that has *much* nicer properties but trades off the ability to introspect.

https://github.com/Atemu/nixpkgs/pull/1

Atemu commented:
Hm, I have also realised that their approach trades off the ability to insert arbitrary custom constraints too as all constraints must be wrapped inside of a function that takes the platform as a parameter.

This is an admittedly fringe case and could be remedied using a simple helper function around the custom constraint that applies the same wrapping.

PerchunPak commented on lib/tests/misc.nix:
```diff
@@ -139,7 +139,7 @@ let
 in
 
-runTests {
+runTests ({
```
Redundant parenthesis?

Atemu replied:
Not redundant.

PerchunPak commented on lib/platform-constraints.nix:
```diff
@@ -0,0 +1,70 @@
+    platform: initialValue:
+    let
+      operations = {
+        OR = lib.any lib.id;
+        AND = lib.all lib.id;
```
(I am not a native, so feel free to ignore this comment)

Wouldn't this naming be better?

```suggestion
        ANY = lib.any lib.id;
        ALL = lib.all lib.id;
```

Atemu replied:
No, the purpose of this attrset is to map between the logical functions and their implementations.

Atemu commented:
I'm far from an expert in interpreting these stats but the [eval perf report](https://github.com/NixOS/nixpkgs/pull/348238/checks?check_run_id=31465324439) appears to show a slight increase in function activity and thunks but a slight decrease in GC heap size?

Anyhow, that's with the functionality not really in use by any packages yet. Real-world testing would likely prove to be difficult to realise. Perhaps we could migrate a critical mass of packages via automation somehow and extrapolate the effect.

adisbladis committed:
lib.platform: optimise platform constraint algebra

This originally had some more opinionated changes and was based on an older
state of my patches. I only kept the performance optimisations.

I was not able to measure a significant change in perf but these look like
they'd be better for perf to me and also improve the code.

Co-authored-by: Atemu <git@atemu.net>

Atemu committed:
lib.platform: expose toPretty customPrinters

Atemu commented:
I did a quick and dirty hack by sed-patching the most common trivial cases of all, unix, linux++darwin, linux and darwin.

Merge base:

```
$ NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 hyperfine --warmup 10 -- "nix-instantiate nixos/release.nix -A iso_minimal.x86_64-linux"
Benchmark 1: nix-instantiate nixos/release.nix -A iso_minimal.x86_64-linux
  Time (mean ± σ):     13.437 s ±  0.119 s    [User: 5.799 s, System: 2.197 s]
  Range (min … max):   13.283 s … 13.706 s    10 runs
```

My PR with 12902 packages hacked:

```
$ NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 hyperfine --warmup 10 -- "nix-instantiate nixos/release.nix -A iso_minimal.x86_64-linux"
Benchmark 1: nix-instantiate nixos/release.nix -A iso_minimal.x86_64-linux
  Time (mean ± σ):     13.465 s ±  0.119 s    [User: 5.823 s, System: 2.187 s]
  Range (min … max):   13.293 s … 13.667 s    10 runs
```

(also: wth, 13s eval for the minimal ISO‽ It was 6s on the previous merge base..)

Sed script used:

```
s/platforms =\s*(with (lib)?(platforms;\s*))?(\S+all);/platforms = lib.platforms.all; platformConstraints = with lib.meta.platform.constraints; ANY;/g
s/platforms =\s*(with (lib)?(platforms;\s*))?(\S+linux);/platforms = lib.platforms.all; platformConstraints = with lib.meta.platform.constraints; isLinux;/g
s/platforms =\s*(with (lib)?(platforms;\s*))?(\S+darwin);/platforms = lib.platforms.all; platformConstraints = with lib.meta.platform.constraints; isDarwin;/g
s/platforms =\s*(with (lib)?(platforms;\s*))?(\S+unix);/platforms = lib.platforms.all; platformConstraints = with lib.meta.platform.constraints; isUnix;/g
s/platforms = (\S+linux)\s*\+\+\s*(\S+darwin);/platforms = lib.platforms.all; platformConstraints = with lib.meta.platform.constraints; OR [ isLinux isDarwin ];/g
s/platforms = (\S+darwin)\s*\+\+\s*(\S+linux);/platforms = lib.platforms.all; platformConstraints = with lib.meta.platform.constraints; OR [ isLinux isDarwin ];/g
```

cmdline to hack:

```
fd --extension nix . pkgs/ | xargs -P 10 sed -i -E -f 'constraints.sed' 
```

Atemu committed:
lib.generators: add customPrinters support to toPretty

Atemu committed:
check-meta: pretty-print platformConstraints errors

Atemu commented:
In talking with @llakala about the current eval perf of the platforms check, I had the idea that we could declare platforms as a function like this:

```
platform = constraints: with constraints; OR [ isLinux isDarwin ]
```

This declares the same meaning but is no longer dependent on the specific definitions in lib, so it could be substituted with a set of pre-computed values for e.g. `isLinux` instead which might be *much* faster to eval. Of course you could also still plug in the constraints values from lib and work with them as data.
